test/abe/dippe.c

```diff
@@ -0,0 +1,214 @@
+    cfe_dippe_pub_key *pks[] = {&pk0, &pk0, &pk0, &pk0, &pk0};
+
+    // unique GID
+    char gid[] = "TESTGID";
+
+    // Test patterns
+    char attr_patterns[6][5] = {
+        "1100", "1001", "0101", "0111", // valid patterns
+        "1101", "1111"                  // invalid patterns
+    };
+    // Test results
+    int decryption_results[] = {1, 1, 1, 1, 0, 0};
+
+    // Policy vector (one additional component is added)
+    //   ------            -----
+    //   Policy            Attrs
+    //   ------            -----
+    //    1                 1
+    //    1                 1
+    //    0         .       0      =  (1+1+0+0-t) = 0
+    //    1                 0
+    //   -t                 1
+    cfe_vec pol;
+    err = cfe_dippe_build_exact_threshold_policy_vector(&dippe, &pol, "1101", 2);
+    munit_assert(err == CFE_ERR_NONE);
+
+    mpz_t     attrs;
+    cfe_dippe_cipher cipher;
+
+    // Decrypt message
+    FP12_BN254 result;
+
+    // Collection of user secret keys
+    cfe_dippe_user_secret_key usks[5];
+
+    for (int i=0; i<6; i++) {
+        // Attribute vector (one additional component is added)
+        err = cfe_dippe_build_exact_threshold_attribute_vector(&dippe, &attrs, attr_patterns[i]);
+        munit_assert(err == CFE_ERR_NONE);
+
+        // Encrypt message under given policy vector
+        err = cfe_dippe_encrypt(&dippe, &cipher, pks, (sizeof(pks)/sizeof(cfe_dippe_pub_key*)), &pol, &msg);
+        munit_assert(err == CFE_ERR_NONE);
+
+        // User secret keys from Auth0
+        for (int j=0; j<5; j++) {
+            err = cfe_dippe_keygen(&dippe, &(usks[j]), j, pks, (sizeof(pks)/sizeof(cfe_dippe_pub_key*)), &sk0, &attrs, gid);
+            munit_assert(err == CFE_ERR_NONE);
+        }
+
+        // Decrypt message
+        err = cfe_dippe_decrypt(&dippe, &result, (cfe_dippe_user_secret_key*)&usks, (sizeof(usks)/sizeof(cfe_dippe_user_secret_key)), &cipher, &attrs, gid);
+        munit_assert(err == CFE_ERR_NONE);
+
+        // Check decryption
+        munit_assert(FP12_BN254_equals(&msg, &result) == decryption_results[i]);
+
+        // Cleanup
+        mpz_clear(attrs);
+        cfe_dippe_cipher_free(&cipher);
+        for (int j=0; j<5; j++) {
+            cfe_dippe_user_secret_key_free(&(usks[j]));
+        }
+    }
+
+    // Cleanup
+    cfe_vec_free(&pol);
+    cfe_dippe_sec_key_free(&sk0);
+    cfe_dippe_pub_key_free(&pk0);
+    cfe_dippe_free(&dippe);
+
+    return MUNIT_OK;
+}
+
+MunitTest dippe_tests[] = {
+    {(char *) "/end-to-end-conjunction",    test_dippe_end_to_end_conjunction,  NULL, NULL, MUNIT_TEST_OPTION_NONE, NULL},
+    {(char *) "/end-to-end-threshold",      test_dippe_end_to_end_threshold,    NULL, NULL, MUNIT_TEST_OPTION_NONE, NULL},
+    {NULL, NULL,                                                                NULL, NULL, MUNIT_TEST_OPTION_NONE, NULL}
+};
+
+MunitSuite dippe_suite = {
+    (char *) "/abe/dippe", dippe_tests, NULL, 1, MUNIT_SUITE_OPTION_NONE
+};
```

Comment:
Have you tried running these tests with [Valgrind](http://valgrind.org) to find any potential memory leaks?

Reply:
I tried running Valgrind's memcheck tool. It didn't report me any memory leaks for the implementation. However, I got some reports but for all defined tests. It seems to be the same as described [here](https://github.com/nemequ/munit/issues/57), so maybe its no real problem.